posts/2022/01/ipython-8.0-lessons-learned-maintaining-software.md

```diff
@@ -0,0 +1,303 @@
+go back to my previous task. 
+
+
+### Be clear what deprecated means, and what the replacements are
+
+While "deprecated" conveys the idea that something should not be used,
+there might be reasons why, or change in effects since the deprecation that
+you want to describe.
+
+
+```python
+def publish_display_data(data, metadata=None, source=None, *, transient=None, **kwargs):
+    """
+    ...
+
+    source : Deprecated.
+
+    """
+    ...
+```
+
+In this particular case the `source=` parameter is deprecated, and passing it
+has, **and had** no effects. It was originally left in the function signature to
+avoid `TypeError`s.
+
+The deprecation without explanation, and suggestion of alternative lead to
+numerous confusion by users (and IPython developers later on). It also delayed
+removal of passing the `source=` arguments in many dependees, and thus cleanup
+of the API.
+
+This is one of the deprecated API we did not remove in 8.0 even if the
+deprecation enacted a couple of years ago, and for 8.0 we've updated the
+deprecation message. A better one if
+
+```python
+    ...
+    warnings.warn("The "source" parameter has been deprecated, it has no effects and can safely omitted, there are no replacements.", DeprecationWarning, stacklevel=2)
+```
+
+If there are replacement, or an option was deprecated as it was obviously wrong,
+you may want to say that as well. In particular if the alternative option are
+available before the deprecation as that can avoid conditional code. 
+
+
+### Indicate the time of the deprecation
+
+This one if a bit of a pet peeve of mine, I regularly come across a deprecation
+and need to go hunt into `git blame` to figure out which versions are affected. 
+Sometime it is written in the function docstring, but still this can interrupt
+my workflow as the DeprecationWarning could be in CI and I don't have the
+library installed locally. 
+
+The version since a deprecation is critical as:
+  - it gives me the right info for a conditional
+  - it tells me whether I can maybe drop support for older version.
+  - It give me an idea of the timeframe for me to fix the deprecation. 
+  - Sometime the warning is added in the different version than the deprecation.
+
+
+The warning from the previous section should become:
+
+```python
+    ...
+    warnings.warn("The "source" parameter has been deprecated since IPython 5.0, "
+                  "it has no effects and can safely omitted, there are no replacements.",
+                  DeprecationWarning, 
+                  stacklevel=2)
+```
+
+
+Some authors also like to add in the deprecation the version where the
+functionality will be removed. I tend to avoid as:
+
+ - I believe this gives users explicit authorisation to delay updating API.
+   Though you as a maintainer want to get rid of it as fast as possible.
+ - I often came across deprecation that were reverted/delayed/not removed. This
+   becomes  confusing to users, and can lead to mistrust of deprecation warning. 
```

Comment:
These are interesting points for not including the version the functionality will be removed.